File: go/tree/1261. 在受污染的二叉树中查找元素.go
package tree

type FindElements struct {
	res  map[int]struct{}
	root *TreeNode
}

func Constructor(root *TreeNode) FindElements {
	f := FindElements{
		res: make(map[int]struct{}),
	}
	var dfs func(*TreeNode, int)
	dfs = func(root *TreeNode, i int) {
		if root == nil {
			return
		}
		root.Val = i
		f.res[root.Val] = struct{}{}
		dfs(root.Left, 2*i+1)
		dfs(root.Right, 2*i+2)
	}
	dfs(root, 0)
	f.root = root
	return f
}

func (this *FindElements) Find(target int) bool {
	if _, ok := this.res[target]; ok {
		return true
	}
	return false
}

File: python/树/1261. 在受污染的二叉树中查找元素.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class FindElements:

    def __init__(self, root: TreeNode):
        self.res = []
        self.dfs(root, 0)
        self.root = root

    def find(self, target: int) -> bool:
        return target in self.res

    def dfs(self, root, v):
        if not root: return
        root.val = v
        self.res.append(root.val)
        if root.left:
            self.dfs(root.left, 2 * root.val + 1)
        if root.right:
            self.dfs(root.right, 2 * root.val + 2)